code/go/0chain.net/chaincore/chain/protocol_block.go

```diff
@@ -345,6 +346,13 @@ func (c *Chain) finalizeBlock(ctx context.Context, fb *block.Block, bsh BlockSta
 	c.SetLatestOwnFinalizedBlockRound(fb.Round)
 	c.SetLatestFinalizedBlock(fb)
 
+	if len(fb.Events) > 0 && c.GetEventDb() != nil {
+		go func(events []event.Event) {
+			c.GetEventDb().AddEvents(ctx, events)
+		}(fb.Events)
+		fb.Events = nil
```

Comment:
I think this might be better synchronous with a timeout. 
```go
eventDb := &EventDb{
  Store:         db,
  eventsChannel: make(chan events, 1000000),
}

func (edb *EventDb) AddEvents(ctx context.Context, events []Event) {
	edb.eventsChannel <- events
}
```
If the timeout fails the event database is set to be offline and can be repaired later. The difference in doing it asynchronously is that it always works even if it doesn't.

If the event database slows down, then the events for a block could be entered out of order leading to a silently corrupt database rather than an offline database. 

We could also add a zero to the `eventsChannel` size to ease the problem if it becomes one.

Reply:
Since the blocks are processed in finalizing process, so actually they have already ordered, and because blocks will be queued in the `eventsChannel`, so it should be fine to do so. 

If we do it in synchronous way, I'm afraid the finalize process would be affect a lot(it's already slow), then the whole chain speed would slow down.

Reply:
If there is plenty of space on the `eventsChannel` there will likely be no problem either way. As pushing the events to the channel should be quick. The problem comes when the `eventsChannel` gets full up,

The scenario I envisage is the `eventsChannel` getting blocked so the AddEvents goroutines are building up,  all trying to add to their events to the channel. When the `eventsChannel` starts to clear the order that blocks of events get added will be undefined.

Maybe we are ok just to assume it won't happen, and if it does add an extra zero to eventsChannle size - which should be a setting not hardcoded.